server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeDataLoader.java

```diff
@@ -0,0 +1,390 @@
+    List<WsMeasures.Period> periods = new ArrayList<>();
+    for (int periodIndex = 1; periodIndex <= 5; periodIndex++) {
+      if (baseSnapshot.getPeriodDate(periodIndex) != null) {
+        periods.add(snapshotDtoToWsPeriod(baseSnapshot, periodIndex));
+      }
+    }
+
+    return periods;
+  }
+
+  private static WsMeasures.Period snapshotDtoToWsPeriod(SnapshotDto snapshot, int periodIndex) {
+    WsMeasures.Period.Builder period = WsMeasures.Period.newBuilder();
+    period.setIndex(periodIndex);
+    if (snapshot.getPeriodMode(periodIndex) != null) {
+      period.setMode(snapshot.getPeriodMode(periodIndex));
+    }
+    if (snapshot.getPeriodModeParameter(periodIndex) != null) {
+      period.setParameter(snapshot.getPeriodModeParameter(periodIndex));
+    }
+    if (snapshot.getPeriodDate(periodIndex) != null) {
+      period.setDate(formatDateTime(snapshot.getPeriodDate(periodIndex)));
+    }
+
+    return period.build();
+  }
+
+  /**
+   * Conditions for best value measure:
+   * <ul>
+   *   <li>component is a production file or test file</li>
+   *   <li>metric is optimized for best value</li>
+   * </ul>
+   */
+  private static void addBestValuesToMeasures(Table<String, MetricDto, MeasureDto> measuresByComponentUuidAndMetric, ComponentTreeData.Builder data) {
+    List<ComponentDtoWithSnapshotId> componentsEligibleForBestValue = from(data.getComponentsFromDb()).filter(IsFileComponent.INSTANCE).toList();
+    List<MetricDtoWithBestValue> metricDtosWithBestValueMeasure = from(data.getMetrics())
+      .filter(IsMetricOptimizedForBestValue.INSTANCE)
+      .transform(new MetricDtoToMetricDtoWithBestValue(data.getPeriods()))
+      .toList();
+
+    for (ComponentDtoWithSnapshotId component : componentsEligibleForBestValue) {
+      for (MetricDtoWithBestValue metricWithBestValue : metricDtosWithBestValueMeasure) {
+        if (measuresByComponentUuidAndMetric.get(component.uuid(), metricWithBestValue.metric) == null) {
+          measuresByComponentUuidAndMetric.put(component.uuid(), metricWithBestValue.metric, metricWithBestValue.bestValue);
+        }
+      }
+    }
+  }
+
+  private static void sortComponents(ComponentTreeData.Builder data) {
+    if (data.getWsRequest().getSort().contains(METRIC_SORT)) {
+      data.setComponentsFromDb(ComponentTreeSort.sortComponents(data));
+    }
+  }
+
+  private static void paginateComponents(ComponentTreeData.Builder data) {
+    ComponentTreeWsRequest wsRequest = data.getWsRequest();
+    if (wsRequest.getSort().contains(METRIC_SORT)) {
+      Paging paging = Paging.forPageIndex(wsRequest.getPage())
+        .withPageSize(wsRequest.getPageSize())
+        .andTotal(data.getComponentCount());
+      data.setComponentsFromDb(from(data.getComponentsFromDb())
+        .skip(paging.offset())
+        .limit(paging.pageSize())
+        .toList());
+    }
+  }
+
+  @CheckForNull
+  private List<String> childrenQualifiers(ComponentTreeWsRequest request, String baseQualifier) {
```

Comment:
it could be a set but it's used in the db query afterwards and MyBatis does not handle Set like List in the `<for each>` tag